// @angular-extensions/model/esm2015/model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { map, shareReplay } from 'rxjs/operators';
import * as i0 from "@angular/core";
/**
 * @template T
 */
export class Model {
    /**
     * @param {?} initialData
     * @param {?} immutable
     * @param {?} sharedSubscription
     * @param {?=} clone
     */
    constructor(initialData, immutable, sharedSubscription, clone) {
        this._data = new BehaviorSubject(initialData);
        this.data$ = this._data.asObservable().pipe(map((data) => immutable
            ? clone
                ? clone(data)
                : JSON.parse(JSON.stringify(data))
            : data), sharedSubscription ? shareReplay(1) : map((data) => data));
    }
    /**
     * @return {?}
     */
    get() {
        return this._data.getValue();
    }
    /**
     * @param {?} data
     * @return {?}
     */
    set(data) {
        this._data.next(data);
    }
}
function Model_tsickle_Closure_declarations() {
    /** @type {?} */
    Model.prototype._data;
    /** @type {?} */
    Model.prototype.data$;
}
/**
 * @template T
 */
export class ModelFactory {
    /**
     * @param {?} initialData
     * @return {?}
     */
    create(initialData) {
        return new Model(initialData, true, false);
    }
    /**
     * @param {?} initialData
     * @return {?}
     */
    createMutable(initialData) {
        return new Model(initialData, false, false);
    }
    /**
     * @param {?} initialData
     * @return {?}
     */
    createMutableWithSharedSubscription(initialData) {
        return new Model(initialData, false, true);
    }
    /**
     * @param {?} initialData
     * @param {?} clone
     * @return {?}
     */
    createWithCustomClone(initialData, clone) {
        return new Model(initialData, true, false, clone);
    }
}
ModelFactory.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ ModelFactory.ngInjectableDef = i0.defineInjectable({ factory: function ModelFactory_Factory() { return new ModelFactory(); }, token: ModelFactory, providedIn: "root" });
function ModelFactory_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ModelFactory.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ModelFactory.ctorParameters;
}
