// @angular-extensions/model/esm5/model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { map, shareReplay } from 'rxjs/operators';
import * as i0 from "@angular/core";
/**
 * @template T
 */
var /**
 * @template T
 */
Model = /** @class */ (function () {
    function Model(initialData, immutable, sharedSubscription, clone) {
        this._data = new BehaviorSubject(initialData);
        this.data$ = this._data.asObservable().pipe(map(function (data) {
            return immutable
                ? clone
                    ? clone(data)
                    : JSON.parse(JSON.stringify(data))
                : data;
        }), sharedSubscription ? shareReplay(1) : map(function (data) { return data; }));
    }
    /**
     * @return {?}
     */
    Model.prototype.get = /**
     * @return {?}
     */
    function () {
        return this._data.getValue();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    Model.prototype.set = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this._data.next(data);
    };
    return Model;
}());
/**
 * @template T
 */
export { Model };
function Model_tsickle_Closure_declarations() {
    /** @type {?} */
    Model.prototype._data;
    /** @type {?} */
    Model.prototype.data$;
}
/**
 * @template T
 */
var ModelFactory = /** @class */ (function () {
    function ModelFactory() {
    }
    /**
     * @param {?} initialData
     * @return {?}
     */
    ModelFactory.prototype.create = /**
     * @param {?} initialData
     * @return {?}
     */
    function (initialData) {
        return new Model(initialData, true, false);
    };
    /**
     * @param {?} initialData
     * @return {?}
     */
    ModelFactory.prototype.createMutable = /**
     * @param {?} initialData
     * @return {?}
     */
    function (initialData) {
        return new Model(initialData, false, false);
    };
    /**
     * @param {?} initialData
     * @return {?}
     */
    ModelFactory.prototype.createMutableWithSharedSubscription = /**
     * @param {?} initialData
     * @return {?}
     */
    function (initialData) {
        return new Model(initialData, false, true);
    };
    /**
     * @param {?} initialData
     * @param {?} clone
     * @return {?}
     */
    ModelFactory.prototype.createWithCustomClone = /**
     * @param {?} initialData
     * @param {?} clone
     * @return {?}
     */
    function (initialData, clone) {
        return new Model(initialData, true, false, clone);
    };
    ModelFactory.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ ModelFactory.ngInjectableDef = i0.defineInjectable({ factory: function ModelFactory_Factory() { return new ModelFactory(); }, token: ModelFactory, providedIn: "root" });
    return ModelFactory;
}());
export { ModelFactory };
function ModelFactory_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ModelFactory.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ModelFactory.ctorParameters;
}
